api/handlers/tag_handler.go

```diff
@@ -27,69 +28,143 @@ func (this *TagHandler) Routes() *restful.WebService {
 
 	service.Route(service.GET("/").
 		Filter(basicAuthenticate).
-		To(this.FindTags).
+		To(t.FindTags).
 		Doc("Lists tags, optionally filtered by the query parameters").
-		Param(service.QueryParameter("name", "tag name to find").DataType("string")).
-		Param(service.QueryParameter("type", "target type for the tag").DataType("string").AllowableValues(data.AllowedTagMap())).
+		Param(service.QueryParameter("type", "Require the EntityType field match the specified parameter").DataType("string")).
+		Param(service.QueryParameter("id", "Require the EntityID field match the specified parameter").DataType("string")).
+		Param(service.QueryParameter("fuzz", "Require the prefix of the EntityID field or 'name' tag match the specified parameter").DataType("string")).
+		Param(service.QueryParameter("version", "Require the 'version' tag match the specified parameter. If 'latest' is used, only the latest will be returned").DataType("string")).
+		Param(service.QueryParameter("environment_id", "Require the 'environment_id' tag match the specified parameter").DataType("string")).
 		Returns(200, "OK", []models.EntityWithTags{}))
 
 	service.Route(service.POST("/").
 		Filter(basicAuthenticate).
-		To(this.CreateTag).
+		To(t.CreateTag).
 		Doc("Create a tag for a service, deploy, or environment").
-		Reads(models.EntityTag{}).
-		Returns(http.StatusCreated, "Created", models.EntityTag{}).
+		Reads(models.Tag{}).
+		Returns(http.StatusCreated, "Created", models.Tag{}).
 		Returns(400, "Invalid request", models.ServerError{}).
-		Writes(models.EntityTag{}))
+		Writes(models.Tag{}))
 
 	service.Route(service.DELETE("/").
 		Filter(basicAuthenticate).
-		To(this.DeleteTag).
+		To(t.DeleteTag).
 		Doc("Delete a tag").
-		Reads(models.EntityTag{}).
+		Reads(models.Tag{}).
 		Returns(http.StatusNoContent, "Deleted", nil))
 
 	return service
 }
 
-func (this *TagHandler) FindTags(request *restful.Request, response *restful.Response) {
-	keys := make(map[string]string)
+func (t *TagHandler) FindTags(request *restful.Request, response *restful.Response) {
+	params := make(map[string]string)
 	for key, val := range request.Request.URL.Query() {
-		keys[key] = val[0]
+		params[key] = val[0]
 	}
 
-	result, err := this.TagData.GetTags(keys)
+	var entityType string
+	var entityID string
+	var fuzz string
+	var latestVersion bool
+
+	// break out special filter params so we don't filter
+	// them by tag.Key and tag.Value
+	if val, ok := params["type"]; ok {
```

Comment:
This logic makes a lot more sense to me. Consider a helper function for removing the special params?